File: thekey-view-dialog-fragment/src/main/java/me/thekey/android/view/dialog/LoginDialogCodeGrantAsyncTask.java
package me.thekey.android.view.dialog;

import android.app.Activity;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.RestrictTo;

import me.thekey.android.core.CodeGrantAsyncTask;
import me.thekey.android.core.TheKeyImpl;

import static android.support.annotation.RestrictTo.Scope.LIBRARY;

@RestrictTo(LIBRARY)
public final class LoginDialogCodeGrantAsyncTask extends CodeGrantAsyncTask {
    private final DialogFragmentCompat mDialog;

    public LoginDialogCodeGrantAsyncTask(final DialogFragmentCompat dialog, @NonNull final TheKeyImpl thekey,
                                         @NonNull final String code, @Nullable final String state) {
        super(thekey, null, code, state);
        mDialog = dialog;
    }

    @SuppressWarnings("unchecked")
    @Override
    protected void onPostExecute(final String guid) {
        super.onPostExecute(guid);

        final Activity activity = mDialog.getActivity();
        if (activity instanceof LoginDialogListener) {
            // trigger the correct callback
            if (guid != null) {
                ((LoginDialogListener<DialogFragmentCompat>) activity).onLoginSuccess(mDialog, guid);
            } else {
                ((LoginDialogListener<DialogFragmentCompat>) activity).onLoginFailure(mDialog);
            }
        }

        // close the mDialog if it is still active (added to the activity)
        if (mDialog.isAdded()) {
            mDialog.dismissAllowingStateLoss();
        }
    }
}

File: thekey-view-dialog-fragment/src/main/java/me/thekey/android/view/dialog/LoginDialogCodeGrantAsyncTask.kt
package me.thekey.android.view.dialog

import android.net.Uri
import com.karumi.weak.weak
import me.thekey.android.TheKey
import me.thekey.android.core.CodeGrantAsyncTask
import me.thekey.android.view.fragment.findListener

internal class LoginDialogCodeGrantAsyncTask(dialog: LoginDialogFragment?, thekey: TheKey, dataUri: Uri) :
    CodeGrantAsyncTask(thekey, dataUri) {
    private val dialog: LoginDialogFragment? by weak(dialog)

    override fun onPostExecute(guid: String?) {
        super.onPostExecute(guid)
        dialog?.run {
            findListener<LoginDialogFragment.Listener>()?.let {
                when {
                    guid != null -> it.onLoginSuccess(this, guid)
                    else -> it.onLoginFailure(this)
                }
            }

            // close the dialog if it is still active
            if (isAdded) dismissAllowingStateLoss()
        }
    }
}
